### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by lavsasha**

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+set(CMAKE_CXX_STANDARD 17)
+
+find_package(MNIST PATHS mnist/)
+
+if(NOT CMAKE_BUILD_TYPE)
+    set(CMAKE_BUILD_TYPE Release)
+endif()
+
+set(CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS_RELEASE} -O3 -DNDEBUG")
```
Comment (DimaTrushin) [inline CMakeLists.txt]: Не пиши компиляторо зависимые флаги без проверки на компилятор. Заверни в
```cpp
if ((CMAKE_CXX_COMPILER_ID MATCHES "Clang") OR (CMAKE_CXX_COMPILER_ID STREQUAL "GNU"))
  ...
endif()
```

File: CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,33 @@
+cmake_minimum_required(VERSION 3.22)
+project(network_project)
+
+set(CMAKE_CXX_STANDARD 17)
+
```
Comment (DimaTrushin) [inline CMakeLists.txt:6]: Добавь строчку
```cpp
set(CMAKE_CXX_STANDARD_REQUIRED True)
```
А то cmake разрешено откатиться на более низкий стандарт.

File: CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+target_include_directories(
+        network_test PUBLIC ${PROJECT_SOURCE_DIR})
+add_executable(test_mnist test_mnist.cpp)
+target_include_directories(
+        test_mnist PUBLIC ${PROJECT_SOURCE_DIR})
+
+target_link_libraries(network_test PUBLIC NeuralNets_lib)
+target_link_libraries(test_mnist PUBLIC NeuralNets_lib)
+
+target_compile_definitions(test_mnist PRIVATE MNIST_DATA_LOCATION="${MNIST_DATA_DIR}")
```
Comment (DimaTrushin) [inline CMakeLists.txt]: Добавь пустую строку в конце каждого файла. Дело в том, что гит на это ругается и пользователь твоего репозитория будет видеть сообщения об ошибках от гита. Не делай так.
Добавить в clang-format правило:
InsertNewlineAtEOF: true

File: main.cpp [hunk 4]
```diff
@@ -0,0 +1,6 @@
+#include "tests.h"
+
+int main() {
```
Comment (DimaTrushin) [inline main.cpp:4]: У тебя не отключены exception, потому надо поставить блок на них. Это делается так
```cpp
#include "tests.h"
#include "Except.h"

int main() {
  try {
    test::RunAllTests();
  } catch(...) {
    except::react();
  }
  return 0;
}
Except.h
namespace except {
void react();
}
Except.cpp
#include "Except.h"

namespace except {
void react() {
  try {
    throw;
  } catch(std::exception& e) {
    // обрабатываешь известные исключения
  } catch(...) {
    // обрабатываешь незивестные исключения
  }
}
}
```

File: usings.h [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+    using IndexType3 = bool;
+    using Index = Eigen::Index;
+    using FunctionType = std::function<Matrix(const Matrix &)>;
+    using PenaltyFuncType = std::function<IndexType2(const Matrix &, const Matrix &)>;
+    using GradientFuncType = std::function<Matrix(const Matrix &, const Matrix &)>;
+
+    enum {
+        measure = 14,
+        sample_num = 5
+    };
```
Comment (DimaTrushin) [inline usings.h]: Лучше не создавай enum. Если хочешь константы во время компиляции можешь сделать constexpr переменную нужного типа. Только снабди глобальные константы префиксом g или g_, то есть gMeasure или g_measure в зависимости от стиля. Чтобы можно было отличать глобальные константы от локальных данных.

File: usings.h [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+
+    static Row GetRowOf1(IndexType1 cols) {
```
Comment (DimaTrushin) [inline usings.h]: Не делай статические функции в хедере, тогда уж inline. Но улчше эти функции переместить туда, где они нужны. Неужели эти функции используются повсеместно?

File: usings.h [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+#include <numeric>
+#include <ranges>
+#include <functional>
+#include <Eigen/Dense>
+
+namespace NeuralNets {
+    using Matrix = Eigen::MatrixXd;
+    using Vector = Eigen::VectorXd;
+    using Row = Eigen::RowVectorXd;
+    using IndexType1 = long long;
```
Comment (DimaTrushin) [inline usings.h]: Не очень правильная логика у тебя тут. Если ты делаешь для удобства using-ги для Eigen имен, это НЕ значит, что надо делать файл со всеми юзингами на свете. Просто смысл делать using с Eigen именами -- это просто создание прокси хедера, когда ты как бы подключаешь Eigen но с дополнительными фичами. Потому я бы назватл этот файл как EigenProxy или EigenExt и в нем бы оставил только using-ги для Eigen. А все остальные юзинги я бы перенесь туда, где они нужны. Вряд ли у тебя все три типа индексов используются везде во всех точках кода. Вот про функции я так вообще уверен, что это нужно либо внутри функции активации либо внутри функции штрафа и больше нигде, потому эти юзинги надо перенести туда.

File: usings.h [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+    using IndexType2 = double;
+    using IndexType3 = bool;
+    using Index = Eigen::Index;
+    using FunctionType = std::function<Matrix(const Matrix &)>;
+    using PenaltyFuncType = std::function<IndexType2(const Matrix &, const Matrix &)>;
+    using GradientFuncType = std::function<Matrix(const Matrix &, const Matrix &)>;
+
+    enum {
```
Comment (DimaTrushin) [inline usings.h]: Не уверен, что тебе нужны эти константы глобально. Перенеси туда, где они используются.

File: Rand.h [hunk 9]
```diff
@@ -0,0 +1,11 @@
+#pragma once
+
+#include "usings.h"
+
+namespace NeuralNets
+{
+    class Rand {
+    public:
+        static Matrix GetNormal(IndexType1 rows, IndexType1 cols);
```
Comment (DimaTrushin) [inline Rand.h]: Плохое название типов. Лучше используй для размеров матриц Eigen::Index. И сделай его глобальным юзингом.

File: ActivationFunction.h [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+#include "usings.h"
+
+namespace NeuralNets {
+
+    enum class AF_id {
+        Sigmoid, ReLu, Tanh
+    };
+
+    template<AF_id>
+    static IndexType2 Evaluate_0(IndexType2);
```
Comment (DimaTrushin) [inline ActivationFunction.h]: для того чтобы понять как действует эта функция надо искать определение IndexType2. Кроме того, у тебя очень плохое название IndexType2 вообще ничего полезного не говорит о данных. Что мешает просто писать double? Или если хочешь иметь какой-то тип, который можно настраивать, то заведи using Scalar = double; и используй его везде. Это хоть как-то будет понятно из названия что делает.

File: ActivationFunction.h [hunk 11]
```diff
@@ -0,0 +1,116 @@
+#pragma once
+
+#include "usings.h"
+
+namespace NeuralNets {
+
+    enum class AF_id {
+        Sigmoid, ReLu, Tanh
+    };
+
```
Comment (DimaTrushin) [inline ActivationFunction.h:11]: Я бы обернул эту базу данных функций шаблонных в дополнительный namespace, чтобы пользователю не вываливать все эти имена в доступ.

File: ActivationFunction.h [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+    template<AF_id>
+    static Matrix Evaluate_0_mat(const Matrix &);
+
+    template<AF_id>
+    static Matrix Evaluate_1_mat(const Matrix &);
+
+    template<>
+    inline IndexType2 Evaluate_0<AF_id::Sigmoid>(IndexType2 x) {
+        IndexType2 result = 1 / (1 + exp(-x));
+        assert(std::isfinite(result));
```
Comment (DimaTrushin) [inline ActivationFunction.h:27]: Круто, Не знал о такой функции.

File: ActivationFunction.h [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+    }
+
+    template<>
+    inline IndexType2 Evaluate_0<AF_id::ReLu>(IndexType2 x) {
+        return x * (x > 0);
+    }
+
+    template<>
+    inline IndexType2 Evaluate_1<AF_id::ReLu>(IndexType2 x) {
+        return x > 0 ? 1 : 0;
```
Comment (DimaTrushin) [inline ActivationFunction.h]: не консистентно, Ты выше пишешь return x * (x > 0);, почему здесь тогда не return x > 0;? Просто соблюдай единый стиль.

File: ActivationFunction.h [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+    }
+
+    template<>
+    inline Matrix Evaluate_1_mat<AF_id::Tanh>(const Matrix &batch) {
+        return batch.unaryExpr(
+                [](IndexType2 x) { return Evaluate_1<AF_id::Tanh>(x); });
+    }
+
+    class ActivationFunction {
+        FunctionType Evaluate_0_;
```
Comment (DimaTrushin) [inline ActivationFunction.h:102]: Перенеси определение FunctionType сюда из глобальных. Оно только тут и нужно и его будет проще найти и понять этот код.

File: LearningRate.cpp
Comment (DimaTrushin) [inline LearningRate.cpp:1]: Очень нехватает clang-format

File: Penalty.h [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+    }
+
+    class Penalty {
+        PenaltyFuncType PenaltyFunc_;
+        GradientFuncType GradientFunc_;
+
+    public:
+        Penalty(PenaltyFuncType PenaltyFunc, GradientFuncType GradientFunc);
+
+        template<PenaltyId penalty> static Penalty Initialize();
```
Comment (DimaTrushin) [inline Penalty.h]: шаблонную функцию надо определять тут же при объявлении, иначе у тебя будет ошибка линковки.

File: test_mnist.cpp [hunk 16]
```diff
@@ -0,0 +1,38 @@
+#include "Network.h"
+#include "mnist/include/mnist/mnist_reader.hpp"
+
+using namespace NeuralNets;
+using std::vector;
+
+Matrix MnistDataX(const vector<vector<unsigned char>> &data) {
```
Comment (DimaTrushin) [inline test_mnist.cpp]: Все что пишется в cpp файлах должно быть в anonymous namespace

File: test_mnist.cpp [hunk 17]
```diff
@@ -0,0 +1,38 @@
...
+    mnist::MNIST_dataset<std::vector, std::vector<uint8_t>, uint8_t> data =
+            mnist::read_dataset<std::vector, std::vector, uint8_t, uint8_t>(MNIST_DATA_LOCATION);
+
+    Network net({784, 256, 10}, {AF_id::ReLu, AF_id::Sigmoid});
+
+    Matrix data_x = MnistDataX(data.training_images);
+    Matrix data_y = MnistDataY(data.training_labels);
+    LearningRate lr(50.0);
+    net.Train(data_x, data_y, 30, Penalty::Initialize(PenaltyId::MSE),
+              0.3, lr, 70);
```
Comment (DimaTrushin) [inline test_mnist.cpp]: Не используй волшебные константы, заведи переменные с соответствующими именами и присвой им эти значения. А тут используй переменные, тогда их имена скажут о их роли в этой функции.

File: test_mnist.cpp [hunk 18]
```diff
@@ -0,0 +1,38 @@
...
+    Matrix data_y(10, data.size());
+    for (IndexType1 i = 0; i < data.size(); ++i) {
+        for (IndexType1 j = 0; j < 10; ++j) {
+            data_y(j, i) = j == data[i] ? 1 : 0;
+        }
+    }
+    return data_y;
+}
+
+int main() {
```
Comment (DimaTrushin) [inline test_mnist.cpp:29]: Почитай в main.cpp комменатрий про организацию функции main.

File: Network.h [hunk 19]
```diff
@@ -0,0 +1,29 @@
...
+        IndexType2 AveragePenalty(const Matrix &data_x, const Matrix &data_y,
+                                           IndexType1 batch_size, const Penalty &penalty) const;
+        IndexType2 SamplePenalty(const Matrix &data_x, const Matrix &data_y, const Penalty &penalty) const;
+        static IndexType3 CheckTrainProcess(IndexType1 cur_era, IndexType2 cur_penalty, IndexType2& all_penalty,
+                                            IndexType2 eps, LearningRate &learning_rate,
+                                            const std::string &option = "default");
+
+    public:
+        Network(const std::initializer_list<Index> &layer_dim, std::initializer_list<AF_id> af_id);
+        Matrix EvaluateNet(const Matrix &batch) const;
```
Comment (DimaTrushin) [inline Network.h]: Не используй имя Net в названии этого метода, у тебя и так объект является сетью, то есть тебя будет заводение имени 
``cpp
Nework net;
net.EvaluateNet(...);
```
Ничего не мешает сделать метод просто Evaluate и он не станет менее понятным.

File: Network.cpp [hunk 20]
```diff
@@ -0,0 +1,127 @@
...
+    }
+
+    void Network::Train(const Matrix &data_x, const Matrix &data_y, IndexType1 batch_size, const Penalty &penalty,
+                        IndexType2 eps, LearningRate &learning_rate, IndexType1 max_era) {
+        IndexType1 check_frequency = max_era / measure > 1 ? max_era / measure : 1;
+        IndexType2 sample_penalty = SamplePenalty(data_x, data_y, penalty);
+        IndexType2 all_penalty = AveragePenalty(data_x, data_y, batch_size, penalty);
+        std::cout << "Initial penalty: " << all_penalty << '\n';
+        for (IndexType1 cur_era = 1; cur_era <= max_era; ++cur_era) {
+            Era(data_x, data_y, batch_size, penalty, learning_rate, cur_era);
```
Comment (DimaTrushin) [inline Network.cpp]: Плохо, что эта функция не содержит глагола в своем названии. Вот что она делает? Был бы глагол было бы понятно, а так лишь понятно, что что-то связанное с эпохой.

File: Network.cpp [hunk 21]
```diff
@@ -0,0 +1,127 @@
...
+            if (cur_era % check_frequency == 0) {
+                std::cout << "Era num: " << cur_era << '\n';
+                IndexType2 cur_penalty = AveragePenalty(data_x, data_y, batch_size, penalty);
+                if (CheckTrainProcess(cur_era, cur_penalty, all_penalty, eps, learning_rate)) {
+                    break;
+                }
+                IndexType2 cur_smpl_penalty = SamplePenalty(data_x, data_y, penalty);
+                CheckTrainProcess(cur_era, cur_smpl_penalty, sample_penalty, eps,
```
Comment (DimaTrushin) [inline Network.cpp]: А что именно проверяется? Корректность? Условия остановки обучения? из названия плох понятно.

File: Network.cpp [hunk 22]
```diff
@@ -0,0 +1,127 @@
...
+        IndexType1 check_frequency = max_era / measure > 1 ? max_era / measure : 1;
+        IndexType2 sample_penalty = SamplePenalty(data_x, data_y, penalty);
+        IndexType2 all_penalty = AveragePenalty(data_x, data_y, batch_size, penalty);
+        std::cout << "Initial penalty: " << all_penalty << '\n';
+        for (IndexType1 cur_era = 1; cur_era <= max_era; ++cur_era) {
+            Era(data_x, data_y, batch_size, penalty, learning_rate, cur_era);
+            if (cur_era % check_frequency == 0) {
+                std::cout << "Era num: " << cur_era << '\n';
+                IndexType2 cur_penalty = AveragePenalty(data_x, data_y, batch_size, penalty);
+                if (CheckTrainProcess(cur_era, cur_penalty, all_penalty, eps, learning_rate)) {
```
Comment (DimaTrushin) [inline Network.cpp]: Плохое название. Если функция возвращает bool, то ее название должно начинаться либо с глагола is/are или с has/have. Потому что не понятно какое именно условие тут проверяется.

File: Network.cpp [hunk 23]
```diff
@@ -0,0 +1,127 @@
...
+        Matrix sample_x = data_x.block(0, 0,
+                                       data_x.rows(), data_x.cols() / sample_num + 1);
+        Matrix sample_y = data_y.block(0, 0,
+                                       data_y.rows(), data_y.cols() / sample_num + 1);
+        return penalty.CalcPenalty(EvaluateNet(sample_x), sample_y);
+    }
+
+    bool Network::CheckTrainProcess(IndexType1 cur_era, IndexType2 cur_penalty, IndexType2& all_penalty, IndexType2 eps,
+                                    LearningRate &learning_rate, const std::string &option) {
+        if (option == "default") {
```
Comment (DimaTrushin) [inline Network.cpp]: Я предпочитаю делать в качестве флагов enum class-ы с нормальными названиями.

File: Network.cpp [hunk 24]
```diff
@@ -0,0 +1,127 @@
...
+            std::cout << "On the " << cur_era << " era penalty is " << cur_penalty << '\n';
+            if (cur_penalty < eps) {
+                std::cout << "The required accuracy has been achieved!\n";
```
Comment (DimaTrushin) [inline Network.cpp]: это только не accuracy, это величина штрафа.

File: Network.cpp [hunk 25]
```diff
@@ -0,0 +1,127 @@
...
+        layer_values.emplace_back(batch);
+        bool is_empty = true;
+        for (const auto &layer: layers_) {
+            if (is_empty) {
+                layer_values.emplace_back(layer.Evaluate(batch));
+                is_empty = false;
+            } else {
+                layer_values.emplace_back(layer.Evaluate(layer_values.back()));
+            }
+        }
```
Comment (DimaTrushin) [inline Network.cpp]: Ну это какая-то дичь. Тебе ненужен этот флаг, потому что ты уже батч положила в самое начало, потому ты можешь сразу написать
```cpp
layer_values.emplace_back(batch);
for (const auto &layer: layers_) {
  layer_values.emplace_back(layer.Evaluate(layer_values.back()));
}
```
И еще в таком виде тут нет разницы использовать emplace_back или push_back.

**Review by DimaTrushin — CHANGES_REQUESTED** [review]
Я сделал несколько мелких замечаний.

Из основного -- добавь clang-format и настрой свою ide так, чтобы она применяла его к файлам при сохранении автоматически. Тогда тебе не придется мучиться с форматированием никогда.

Еще бы хотелось видеть clang-tidy.

Еще я бы разнес usings.h на разные части. Я написал подробно что мне там не нравится. Ты по сути смешала слишком много разного и поместила в него. Часть из этих юзингов нужны ближе к их использованию, иначе тяжело в коде искать где же это все определено.

Есть несколько мест где код криво написан, но это все локальные мелочи.

File: usings.h (hunk 6 above)
Comment (lavsasha) [reply usings.h:29]: они используются в двух классах
